enterprise/backend/src/metabase_enterprise/transforms_python/python_runner.clj

```diff
@@ -0,0 +1,321 @@
+    ;; when a 500 is returned we observe a string in the body (despite the python returning json)
+    ;; always try to parse the returned string as json before yielding (could tighten this up at some point)
+    (update response :body (fn [string-if-error]
+                             (if (string? string-if-error)
+                               (try
+                                 (json/decode+kw string-if-error)
+                                 (catch Exception _
+                                   {:error string-if-error}))
+                               string-if-error)))))
+
+(defn- cancel-python-code-http-call! [server-url run-id]
+  (python-runner-request server-url :post "/cancel" {:body   (json/encode {:request_id run-id})
+                                                     :async? true}
+                         #_success #(log/debug %)
+                         #_failure #(log/error %)))
+
+(defn open-cancellation-process!
+  "Starts a core.async process that optimistically sends a cancellation request to the python executor if cancel-chan receives a value.
+  Returns a channel that will receive either the async http call j.u.c.FutureTask in the case of cancellation, or nil when the cancel-chan is closed."
+  [server-url run-id cancel-chan]
+  (a/go (when (a/<! cancel-chan)
+          (cancel-python-code-http-call! server-url run-id))))
+
+;; temporary, we should not need to realize data/events files into memory longer term
+(defn read-output-objects
+  "Temporary function that strings/jsons stuff in S3 and returns it for compatibility."
+  [{:keys [s3-client bucket-name objects]}]
+  (let [{:keys [output output-manifest events]} objects
+        output-content          (s3/read-to-string s3-client bucket-name (:path output) nil)
+        output-manifest-content (s3/read-to-string s3-client bucket-name (:path output-manifest) "{}")
+        events-content          (s3/read-to-string s3-client bucket-name (:path events))]
+    {:output          output-content
+     :output-manifest (json/decode+kw output-manifest-content)
+     :events          (mapv json/decode+kw (str/split-lines events-content))}))
+
+(defn- safe-delete
+  "Safely delete a file."
+  [^File file]
+  (try (.delete file) (catch Exception _)))
+
+(defn- fields-metadata [_driver table-id]
+  (->> (t2/select [:model/Field :id :name :base_type :effective_type :semantic_type :database_type :database_position :nfc_path :parent_id]
+                  :table_id table-id
+                  :active true
+                  ;; we are only interested in top-level objects, so filter out nested fields (parent or path)
+                  :parent_id nil
+                  :nfc_path nil
+                  {:order-by [[:database_position :asc]]})
+
+       ;; adjust metadata where necessary. disabled for now as we've restricted ourselves only to basic types for v1.
+       ;; TODO introduce a multimethod once we do this in anger
+       identity
+       #_(map (fn [meta]
```

Comment:
should probably delete this code

Reply:
[460e29f](https://github.com/metabase/metabase/pull/63747/commits/460e29f0be85de16183cf00bf0d13a8a3677d98d)